### PR TITLE
Added TryToContinue exception handling to cmsRun

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Removed SkipEvent and FailPath exception handling in favor of TryToContinue. The new handling option will cause a job to continue on event processing if an exception associated to TryToContinue is thrown. For the event which had the exception, 
- If a module is marked as `shouldTryToContinue()` in the configuration, it will be eligible to be run even if a TryToContinue exception happens. Marked modules will also not propagate a TryToContinue exception to the rest of the system whether the exception happened during prefetching or while running the module. If a marked module  is on a Path, it will not propagate the exception to the Path. However, if another module on the Path is dependent on the throwing module and is not marked then the Path will still fail (see next item).
- all Paths with a module that is directly or indirectly dependent on the module which threw will be marked as having an exception. Any modules on the Path which are directly or indirectly dependent upon the throwing module will not be run, unless the dependent module is marked `shouldTryToContinue()`. Any modules on the Path not already scheduled will not be run, including those marked as `shouldTryToContinue()`. Those which were scheduled and independent of the thrown module will still be run.
- Unscheduled modules which depend on the throwing module will not be run, unless they are marked as `shouldTryToContinue()`. Unscheduled modules which are independent may still be run if their data is requested.

#### PR validation:

All the framework unit tests (including the newly added ones) pass.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

**Comment by tvami** [comment]
hi @Dr15Jones could you please squash the commits? I also see a comment "//TODO have to deal with 'cms.catch'" is that for a follow-up PR?

**Comment by Dr15Jones** [comment]
@tvami I'd rather not squash the commits as each has a purpose [except for the code formatting].

As for the `//TODO` for `cms.catch` those changed were already applied and I just failed to notice them before doing the commit. <sigh/>

**Comment by perrotta** [comment]
> As for the `//TODO` for `cms.catch` those changed were already applied and I just failed to notice them before doing the commit.

Then could you please remove them, so that future maintainers will not get confused by them?

**Comment by Dr15Jones** [comment]
> Then could you please remove them, so that future maintainers will not get confused by them?

I will. I'll be doing myself a favor since I am most likely the 'future maintainer' :).

**Comment by emanueleusai** [comment]
are the DQM differences in tracking spurious? the code does not seem to touch any tracking modules beyond alca, but I wonder if they come from events that were previously being skipped now just raising TryToContinue

**Comment by Dr15Jones** [comment]
> are the DQM differences in tracking spurious? the code does not seem to touch any tracking modules beyond alca, but I wonder if they come from events that were previously being skipped now just raising TryToContinue

I double checked and all the RelVals just use the default setting for the exception handling which is to always rethrow any exception seen (i.e. FailPath, SkipEvent and IgnoreCompletely are always empty). Therefore this code change could not be the cause of any spurious change in the DQM.

**Review by makortel — COMMENTED** [review]
Could you add tests also for these cases
- a `TryToContinue` exception is thrown from a non-Event function
- a `shouldTryToContinue()` module throws an `TryToContinue` exception from a non-Event function
- an ESModule throws a `TryToContinue` exception
  - when the job does not have any `shouldTryToContinue()` EDModule
  - when the job has a `shouldTryToContinue()` EDModule that does not consume the ESProduct of the ESModule
  - when the job has a `shouldTryToContinue()` EDModule that consumes the ESProduct of the ESModule
- a Source throws a `TryToContinue` exception (say during Event processing)
- an OutputModule throws a `TryToContinue` exception (say during Event processing, in both cases of `shouldTryToContinue()` and without)
- check that the `TryToContinue` exception message is printed out

Is the current testing sufficient for ensuring the Path with a throwing module is marked in "error" state in `TriggerResults`? (in principle the OutputModule `SelectEvents` uses that, but I'm wondering if a more explicit check would be useful)

According to the plan https://github.com/cms-sw/cmssw/issues/41512#issuecomment-1615284641 all those should result in job termination.

Would the test you did in https://github.com/cms-sw/cmssw/issues/41512#issuecomment-1664127254 be worthwhile to include as unit test?

**Comment by Dr15Jones** [comment]
We are getting another round of comments before making further changes. 

As a reminder to myself, we also determined `TryToContinue` can cause problems if ROOT fast copying is used. (This problem was also missed for the old `SkipEvent` cases). We will want to disable fast copying if `TryToContinue` is in use.

**Comment by Dr15Jones** [comment]
hold

we have opened a conversation in issue about possible alternative user interfaces for setting the `shouldTryToContinue`.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

**Comment by emanueleusai** [comment]
please test

**Comment by Dr15Jones** [comment]
The same test failure is seen in the IB

**Comment by emanueleusai** [comment]
+1

- unrelated uint test failure

**Comment by Dr15Jones** [comment]
unhold

**Comment by tvami** [comment]
@cmsbuild , please test
 * the unit test should be fixed in the IBs now

**Comment by makortel** [comment]
@cmsbuild, please test

File: FWCore/Integration/test/BuildFile.xml [hunk 1]
```diff
@@ -113,10 +113,21 @@
...
+  <test name="TestFWCoreIntegrationTryToContinueLumi" command="cmsRun ${LOCALTOP}/src/FWCore/Integration/test/test_TryToContinue_cfg.py -- --inLumi 2>&amp;1 | fgrep 'exception for testing purposes'"/>
   <test name="TestFWCoreIntegrationShouldTryToContinue" command="cmsRun ${LOCALTOP}/src/FWCore/Integration/test/test_shouldTryToContinue_cfg.py"/>
+  <test name="TestFWCoreIntegrationShouldTryToContinueRun" command="cmsRun ${LOCALTOP}/src/FWCore/Integration/test/test_shouldTryToContinue_cfg.py -- --inRun 2>&amp;1 | fgrep 'exception for testing purposes'"/>
+  <test name="TestFWCoreIntegrationShouldTryToContinueLumi" command="cmsRun ${LOCALTOP}/src/FWCore/Integration/test/test_shouldTryToContinue_cfg.py -- --inLumi 2>&amp;1 | fgrep 'exception for testing purposes'"/>
   <test name="TestFWCoreIntegrationShouldTryToContinueIndirect" command="cmsRun ${LOCALTOP}/src/FWCore/Integration/test/test_shouldTryToContinue_cfg.py -- --indirect"/>
+  <test name="TestFWCoreIntegrationShouldTryToContinueIndirect" command="cmsRun ${LOCALTOP}/src/FWCore/Integration/test/test_shouldTryToContinue_cfg.py -- --indirect"/>
```
Comment (makortel) [inline FWCore/Integration/test/BuildFile.xml:124]: Is this test now being run twice?

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
These unit tests are already failing in IB

**Comment by makortel** [comment]
+core

**Comment by tvami** [comment]
+alca

**Comment by Dr15Jones** [comment]
please test

**Comment by mandrenguyen** [comment]
+1

**Comment by emanueleusai** [comment]
+1

**Comment by rappoccio** [comment]
+1

**Pull request merged**